receiver/dcgmreceiver/scraper.go

```diff
@@ -31,23 +31,45 @@ import (
 )
 
 type dcgmScraper struct {
-	config   *Config
-	settings receiver.CreateSettings
-	client   *dcgmClient
-	mb       *metadata.MetricsBuilder
+	config        *Config
+	settings      receiver.CreateSettings
+	client        *dcgmClient
+	mb            *metadata.MetricsBuilder
+	handleCleanup func()
 }
 
 func newDcgmScraper(config *Config, settings receiver.CreateSettings) *dcgmScraper {
 	return &dcgmScraper{config: config, settings: settings}
 }
 
-func (s *dcgmScraper) start(_ context.Context, _ component.Host) error {
-	var err error
-	s.client, err = newClient(s.config, s.settings.Logger)
+// initClient will try to create a new dcgmClient if currently has no client;
+// it will try to initialize the communication with the DCGM service; if
+// success, create a client; only return errors if DCGM service is available but
+// failed to create client.
+func (s *dcgmScraper) initClient() error {
+	if s.client != nil {
+		return nil
+	}
+	if !isDcgmInstalled() {
+		s.settings.Logger.Warn("can not initialize a DCGM client; DCGM is not installed.")
+		return nil
+	}
+	dcgmCleanup, err := initializeDcgm(s.config, s.settings.Logger)
+	if err != nil {
+		// If can not connect to DCGM, return no error and retry at next
+		// collection time
+		return nil
+	}
+	client, err := newClient(s.config, s.settings.Logger)
 	if err != nil {
 		return err
 	}
+	s.client = client
+	s.handleCleanup = dcgmCleanup
```

Comment:
I'm honestly a bit confused why you can no longer store `handleCleanup` in the client. Is this because `newClient` may be called more than once? But then so would `initializeDcgm`…

Reply:
Done! Moved the `handleCleanup` back to the client. 

Previously, I moved `handleCleanup` up to the scraper as I wanted to call `initializeDcgm()` in the scraper to test if the dcgm can be initialized before create a client. But I agree it can be confusing. 

Right now I've moved the `handleCleanup` back to the client. When calling `newClient()` from the scraper, in order to tell if the error is an initialization error (which can be ignored and retried) or other error, I've defined an error type `ErrDcgmInitialization` and check that.